DatasetBuilder and DatasetBenchmarker

ngreenwald opened the pull request:
This PR creates a DatasetBuilder and DatasetBenchmarker class for constructing and evaluating model accuracy on a range of different datasets. 

**DatasetBuilder**
1. The DatasetBuilder assumes that the data is stored in a folder structure. Each folder corresponds to a single experiment, with a single metadata file. However, there can be multiple NPZ files within each experiment representing different caliban jobs, different days of the same experiment, etc. This will eventually be replaced with a call to deepcell-datasets.
2. The motivation for this class was the lack of metadata that is retained in our current NPZ setup. In particular, we can't map individual images within the array back to their phenotype information. This class extends the dict{} structure we've been using before with two additional keys. tissue_list is a list of the same length as the number of images with the tissue type of each image. platform_list is the same but for image platform.
3.  The trigger for creating a dataset is calling the build_dataset function. However, this method doesn't do much work itself, and is mostly in charge of calling the accessory functions. 
4. `_load_experiment` loads a single folder of data. `_load_all_experiments` loads the entire dataset, and performs the appropriate train/va/test split of the data as it is being loaded in. `_subset_data_dict` takes a dictionary containing X, y, tissue, and platform type, and subsets it to include the specified tissue and platform types. `_reshape_dict` reshapes each of the dicts to account for cropping and resizing of the training data. `_clean_labels` cleans up the labels prior to creating the final dict. 

**DatasetBenchmarker**
This class takes a set of true labels, predicted labels, and a list of the tissue and platform type for each image, and computes metrics across each unique tissue and platform type within the dataset. I was able to abstract this one a bit more, and it's fairly clear how additional categories would be able to easily slot in (i.e. culture condition, resolution, etc). 

**Other changes**
Adds a new function, create_train_val_test_split, which creates the specified split in the data. 
Removes combine_npz_files, which was a filler for the DataLoader

**Areas in need of TLC**
DatasetBuilder._load_all_experiments has 12 different lists [X, y, tissue_list, platform_list] x [train, val, test]. Each of these have a separate _batch version which is used as a working version during the for loop, which then gets appended to the main list, and finally concatenated. The logic is 100% identical between them. What would be a better way to handle this? I was thinking maybe zipping them all together?

DatasetBuilder._reshape_dict has almost the exact same code for reshape_by_img and reshape_by_tissue. The difference is that the first loops over each image, whereas the second loops over each unique tissue type. But the machinery is identical within them. 

build.train_val_test_split currently fails if there aren't enough images to be divided among the three classes. What is the best way to handle this?

ngreenwald commented:
Updated the data_splitting function to handle small datasizes that don't fit into assigned splits

willgraf commented on caliban_toolbox/utils/misc_utils.py:
```diff
@@ -119,3 +119,19 @@ def list_npzs_folder(npz_dir):
+    Returns:
+        list of folders in base_dir, empty if None
+    """
+
+    files = os.listdir(base_dir)
+    folders = [file for file in files if os.path.isdir(os.path.join(base_dir, file))]
+
+    return folders
```
[This might be improved by using `os.walk` instead of `os.listdir`](https://stackoverflow.com/a/973488)

willgraf commented on caliban_toolbox/DatasetBenchmarker.py:
```diff
@@ -0,0 +1,118 @@
+        stats_dict = {}
+        for uid in unique_ids:
+            stats_dict[uid] = {}
+            category_idx = np.isin(category_ids, uid)
```
This seems like its a boolean, not an index? How is this used?

ngreenwald commented:
- Modified train_val_test_split so that all decisions are random, based on seed
- Raises a warning if data split has to be modified
- Adds a validate_dataset function to make sure each folder has an NPZ and metadata file
- Adds a get_metadata function to further abstract where metadata is stored
- combines repeated code into single _reshape_dict loop for resizing

willgraf commented on caliban_toolbox/build.py:
```diff
@@ -43,14 +45,15 @@ def compute_cell_size(npz_file, method='median', by_image=True):
 
     Returns:
-        average_sizes: list of typical cell size in NPZ
+        average_sizes: list of typical cell size in NPZ. If no cells, returns None
```
```suggestion
        list: list of typical cell size in NPZ. If no cells, returns None.
```

The text before the colon is supposed to be the data type, not the var name.

willgraf commented on caliban_toolbox/build.py:
```diff
@@ -43,14 +45,15 @@ def compute_cell_size(npz_file, method='median', by_image=True):
-    valid_methods = set(['median', 'mean'])
-    if method.lower() not in valid_methods:
+    valid_methods = {'median', 'mean'}
+    method = method.lower()
```
You can cast as a string to make sure that `.lower()` doesn't throw an attribute error.

```suggestion
    method = str(method).lower()
```

willgraf commented on caliban_toolbox/dataset_builder.py:
```diff
@@ -0,0 +1,571 @@
+                    resize_ratio = 1
+
+                # resize the data
+                X_batch_resized, y_batch_resized = \
+                    reshape_training_data(X_data=X_batch, y_data=y_batch,
+                                          resize_ratio=resize_ratio, final_size=output_shape,
+                                          tolerance=resize_tolerance)
```
No need for the `\` if the line breaks are inside a paranethesis.

```suggestion
                X_batch_resized, y_batch_resized = reshape_training_data(
                    X_data=X_batch, y_data=y_batch,
                    resize_ratio=resize_ratio, final_size=output_shape,
                    tolerance=resize_tolerance)
```

willgraf commented on caliban_toolbox/dataset_builder.py:
```diff
@@ -0,0 +1,571 @@
+
+        # TODO: remove one data QC happens in main toolbox pipeline
+        for i in range(y.shape[0]):
+            if relabel_hard:
+                y_new = label(y[i, :, :, 0])
+            else:
+                y_new, _, _ = relabel_sequential(y[i, :, :, 0])
```
I haven't done the legwork, but I am still unconvinced this is important.

willgraf commented on caliban_toolbox/dataset_builder.py:
```diff
@@ -0,0 +1,571 @@
+            self._identify_tissue_and_platform_types()
+
+        # validate inputs
+        if isinstance(tissues, list):
+            for tissue in tissues:
+                if tissue not in self.all_tissues:
+                    raise ValueError('{} is not one of {}'.format(tissue, self.all_tissues))
+        elif tissues == 'all':
+            tissues = self.all_tissues
+        elif tissues in self.all_tissues:
+            tissues = [tissues]
+        else:
+            raise ValueError(
+                'tissues should be "all", one of {}, or a list '
+                'of acceptable tissue types'.format(self.all_tissues))
```
These 2 bits of logic look so similar, it could probably be a `validate` function that works for both tissues and platforms.

ngreenwald commented:
Changes:

- Added explicit checks for 4D data where it's currently required, abstracted away where it's not
- Added type checking for labeled data (must be int) for skimage functions to properly work
- Removed relabel_sequential in favor of label, which accomplishes same thing

willgraf commented on caliban_toolbox/build_test.py:
```diff
@@ -167,100 +224,77 @@ def test_pad_image_stack():
-
-    # check that correct number of NPZs present
-    assert combined_x.shape[0] == np.sum(2 + 1 + 4)
+def test_train_val_test_split():
+    X_data = np.zeros((100, 5, 5, 3))
+    y_data = np.zeros((100, 5, 5, 1))
```
You may want to add another test that confirms reproducibility with a given seed.

willgraf commented on caliban_toolbox/dataset_builder_test.py:
```diff
@@ -0,0 +1,615 @@
+import numpy as np
+from pathlib import Path
+
+
```
```suggestion
```

ngreenwald commented:
- Added the option to specify different output sizes for each data split
- Added the option to resize by a constant value for all images, for example to for different resolutions
- Added check for split reproducibility with seed

ngreenwald commented:
Added support for any combination of a list or tuple for various output_shape parameters

willgraf commented on caliban_toolbox/dataset_builder.py:
```diff
@@ -0,0 +1,627 @@
+                    - False. No resizing
+                    - float/int: Resizes all images by supplied value
+                    - by_tissue. Resizes by median cell size within each tissue type
+                    - by_image. REsizes by median cell size within each image
```
```suggestion
                    - by_image. Resizes by median cell size within each image
```

willgraf commented on caliban_toolbox/dataset_builder.py:
```diff
@@ -0,0 +1,627 @@
+        else:
+            X_new, y_new, tissue_list_new, platform_list_new = [], [], [], []
+
+            if resize == 'by_tissue':
+                batch_ids = np.unique(tissue_list)
+            else:
+                batch_ids = np.arange(0, X.shape[0])
```
Even though the input gets validated in a different file, I would feel better if we specifically checked the string value instead of inferring it in the else.

```suggestion
            if resize == 'by_tissue':
                batch_ids = np.unique(tissue_list)
            elif resize == 'by_image':
                batch_ids = np.arange(0, X.shape[0])
            else:
                raise ValueError('Invalid `resize` value: {}'.format(resize))
```

ngreenwald commented:
For sure, always better to be safe

ngreenwald commented on caliban_toolbox/dataset_builder.py:
```diff
@@ -0,0 +1,629 @@
+        keep_idx = np.repeat(True, y.shape[0])
+        cleaned_y = np.zeros_like(y)
+
+        # TODO: remove one data QC happens in main toolbox pipeline
```
```suggestion
        # TODO: remove once data QC happens in main toolbox pipeline
```
